SharkCage/CageServiceInstaller/SharkCageInstallerHelper.cs

```diff
@@ -0,0 +1,143 @@
+﻿using Microsoft.Win32;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections;
+using System.ComponentModel;
+using System.Configuration.Install;
+using System.IO;
+using System.Security.AccessControl;
+using System.Security.Cryptography;
+using System.Security.Principal;
+
+namespace SharkCageInstallerHelper
+{
+    [RunInstaller(true)]
+    public partial class SharkCageInstallerHelper : Installer
+    {
+        const string service_name = "shark-cage-service";
+
+        public override void Install(System.Collections.IDictionary state_saver)
+        {
+            base.Install(state_saver);
+        }
+
+        public override void Commit(System.Collections.IDictionary saved_state)
+        {
+            base.Commit(saved_state);
+
+            string target_dir = Context.Parameters["DP_TargetDir"].ToString();
+
+            AssimilateConfigToCurrentSystem(target_dir);
+
+            InstallAndStartService(target_dir);
+        }
+
+        public override void Rollback(IDictionary saved_state)
+        {
+            base.Rollback(saved_state);
+            UninstallService();
+        }
+
+        public override void Uninstall(System.Collections.IDictionary saved_state)
+        {
+            base.Uninstall(saved_state);
+            UninstallService();
+        }
+
+        private void AssimilateConfigToCurrentSystem(string dir_path)
+        {
+            var config_path = Path.Combine(
+                Environment.GetFolderPath(Environment.SpecialFolder.CommonDocuments),
+                "SharkCage\\CageConfigurator.sconfig");
+
+            Registry.SetValue(@"HKEY_LOCAL_MACHINE\SOFTWARE\SharkCage\Configs", "CageConfigurator", config_path);
+
+            MoveConfigFileToConfigPath(config_path, dir_path);
+
+            UpdateContentOfConfigFile(config_path, dir_path);
+
+            UpdateSecurityAclsOfConfigFile(config_path);
+        }
+
+        private void MoveConfigFileToConfigPath(string config_path, string dir_path)
+        {
+            if (File.Exists(config_path))
+            {
+                File.Delete(config_path);
+            }
+
+            File.Move(Path.Combine(dir_path, "CageConfigurator.sconfig"), config_path);
+        }
+
+        private void UpdateContentOfConfigFile(string config_path, string dir_path)
+        {
+            const string APPLICATION_PATH_PROPERTY = "application_path";
+            const string APPLICATION_HASH_PROPERTY = "binary_hash";
+            var path_to_config_exe = Path.Combine(dir_path, "CageConfigurator.exe");
+
+            var json = JObject.Parse(File.ReadAllText(config_path));
```

Comment:
the json parsing can fail (if the file is missing or has a different structure), do we need a try catch here or does the installer handle it for us?

Reply:
The installer is handling the exception and also will rollback all changes on the system, will display an error message with the content of the exception so nothing to do for us I think...